Add UnitaryOverlap to the circuit library

nonhermitian opened the pull request:
### Summary
Adds the `UnitaryOverlap` circuit to the circuit library that aids in the construction of applications such as machine learning and chemistry.

This circuit returns the fidelity of states generated by two unitaries as the probability of being in the all-zeros state or, equivalently, the expectation value of the projector over the all-zeros state. 


### Details and comments

nonhermitian committed:
add overlap function

nonhermitian committed:
Add tests

nonhermitian committed:
add release note

nonhermitian committed:
add checks for meas+reset

nonhermitian committed:
update docstring

nonhermitian committed:
spelling

Cryoris commented on qiskit/circuit/library/overlap.py:
```diff
@@ -0,0 +1,61 @@
+    $\left|\langle 0\dots 0| V^{\dag}U|0\dots 0\rangle\right|^{2}$ by computing the probability
+    of being in the all-zeros bit-string, or equivalently, the expectation value of projector
+    $|0\dots 0\rangle\langle 0\dots 0|$.
+    """
```
Afaik we have to use the `:math:` environment for TeX to be rendered correctly 🙂 
```suggestion
    r"""Circuit that returns the overlap between two unitaries :math:`V^\dagger U|0\rangle`.

    Input quantum circuits must represent unitary operations that are invertible. Parameters
    in the circuits ``U`` and ``V`` will be replaced by a :class:`.ParameterVector` with name
    ``"a"`` and ``"b"`` in the output circuit.

    A unitary overlap circuit is usually employed in computing the fidelity
    :math:`|\langle 0| V^\dagger U |0\rangle|^2` by computing the probability
    of being in the all-zeros bit-string, or equivalently, the expectation value of projector
    :math:`|0\rangle\langle 0|` acting on all qubits.
    """
```

Cryoris commented on qiskit/circuit/library/overlap.py:
```diff
@@ -0,0 +1,61 @@
+        Raises:
+            CircuitError: Number of qubits in U and V does not match
+            CircuitError: Inputs contain measurements and/or resets
+        """
```
```suggestion
    def __init__(self, U: QuantumCircuit, V: QuantumCircuit) -> None:
        """
        Args:
            U: Unitary acting on the ket vector.
            V: Unitary whose inverse operates on the bra vector.

        Raises:
            CircuitError: Number of qubits in ``U`` and ``V`` does not match.
            CircuitError: Inputs contain non-unitary operations.
        """
```

Cryoris commented on qiskit/circuit/library/overlap.py:
```diff
@@ -0,0 +1,61 @@
+        if any(set(U.count_ops()) & {"measure", "reset"}) or any(
+            set(V.count_ops()) & {"measure", "reset"}
+        ):
+            raise CircuitError("Input circuits contain measurements or resets")
```
There are other operations that would prevent a circuit from being invertible, such as control-flow ops (or any non-`Gate` object). One way to check would be to simply try inverting the circuit and raise an error if that fails. Alternatively, you could iterate over the circuit operations and check if all ops are `Gate` instance.

The first option would be less restrictive and also allow for `U` to be non-unitary (or `V`s where users implement a custom `inverse`). Option 2 would actually ensure that the circuits are unitary from a Qiskit POV

Cryoris commented on qiskit/circuit/library/overlap.py:
```diff
@@ -0,0 +1,61 @@
+        # Generate the actual overlap circuit
+        overlap = circ1.compose(circ2.inverse(), inplace=False)
+        super().__init__(*overlap.qregs, name="UnitaryOverlap")
+        self.compose(overlap, qubits=self.qubits, inplace=True)
```
It might be slightly more efficient to first compose `circ1` onto `self` and then `circ2.inverse` -- that would append  gates(U) and gates(V) once, whereas now it appends 2 times gates(V) and once gates(U).

Cryoris commented on releasenotes/notes/unitary-overlap-7dde87a8e2d7a1f7.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+features:
+  - |
+    Adds the `UnitaryOverlap` circuit to the Qiskit circuit library that
```
```suggestion
    Adds the :class:`.UnitaryOverlap` circuit to the Qiskit circuit library that
```

Cryoris commented on qiskit/circuit/library/overlap.py:
```diff
@@ -0,0 +1,61 @@
+    """
+
+    def __init__(self, U, V):
+        """Create unitary overlap circuit
```
What do you think about having a small example on how to use this circuit in practice?

Cryoris commented on test/python/circuit/library/test_overlap.py:
Could you add one brief test to raise the error of mismatching numbers of qubits? 🙂

Cryoris reviewed (commented):
Overall LGTM, the one thing we have to cover is that there are more ways that circuits can be non-unitary, other than that I just have small comments 🙂

Cryoris committed:
apply review comments

main change: check if circuits are unitary by checking if all operations are of type ``Gate``

Cryoris committed:
Merge branch 'main' into unitary-overlap

Cryoris committed:
Merge branch 'main' into unitary-overlap

Cryoris committed:
Merge branch 'unitary-overlap' of github.com:nonhermitian/qiskit-terra into unitary-overlap

Cryoris committed:
apply changes from review

- do not allow U/V
- allow setting parameter prefixes

jakelishman reviewed (approved):
In principle this looks fine to me now.  It's better that the argument names have a more consistent way of linking to the (default) parameter names, rather than us using three separate U/V, circ1/circ2 and a/b pairs to refer to the same split throughout the code and the public interface.

I'm still of the opinion that these kinds of things should be helper functions and not subclasses of `QuantumCircuit`, but it's better that this new object is consistent with the rest of the circuit library.  The `_check_unitary` helper is slightly overcautious (it will reject directives like `Barrier`, for example), but in lieu of us having a better model for parametric gates at the moment, it's fine.
